cpp/src/io/fst/lookup_tables.cuh

```diff
@@ -0,0 +1,547 @@
+  static constexpr int32_t MAX_NUM_STATES = NUM_STATES;
+
+ private:
+  // Symbol-group id lookup table
+  using SymbolGroupIdLookupT = detail::SingleSymbolSmemLUT<char>;
+  using SymbolGroupIdInitT   = typename SymbolGroupIdLookupT::KernelParameter;
+
+  // Transition table
+  using TransitionTableT     = detail::TransitionTable<NUM_SYMBOLS, NUM_STATES>;
+  using TransitionTableInitT = typename TransitionTableT::KernelParameter;
+
+  // Translation lookup table
+  using OutSymbolOffsetT      = uint32_t;
+  using TranslationTableT     = detail::TransducerLookupTable<OutSymbolT,
+                                                          OutSymbolOffsetT,
+                                                          NUM_SYMBOLS,
+                                                          NUM_STATES,
+                                                          NUM_SYMBOLS * NUM_STATES>;
+  using TranslationTableInitT = typename TranslationTableT::KernelParameter;
+
+  auto get_device_view()
+  {
+    return dfa_device_view<SymbolGroupIdLookupT, TransitionTableT, TranslationTableT, NUM_STATES>{
+      sgid_init.d_begin(), transition_table_init.d_begin(), translation_table_init.d_begin()};
+  }
+
+ public:
+  /**
+   * @brief Constructs a new DFA.
+   *
+   * @param symbol_vec Sequence container of symbol groups. Each symbol group is a sequence
+   * container to symbols within that group. The index of the symbol group containing a symbol being
+   * read will be used as symbol_gid of the transition and translation tables.
+   * @param tt_vec The transition table
+   * @param out_tt_vec The translation table
+   * @param stream The stream to which memory operations and kernels are getting dispatched to
+   */
+  template <typename StateIdT, typename SymbolGroupIdItT>
+  Dfa(SymbolGroupIdItT const& symbol_vec,
+      std::vector<std::vector<StateIdT>> const& tt_vec,
+      std::vector<std::vector<std::vector<OutSymbolT>>> const& out_tt_vec,
+      cudaStream_t stream)
+  {
+    constexpr std::size_t single_item = 1;
+
+    sgid_init              = hostdevice_vector<SymbolGroupIdInitT>{single_item, stream};
+    transition_table_init  = hostdevice_vector<TransitionTableInitT>{single_item, stream};
+    translation_table_init = hostdevice_vector<TranslationTableInitT>{single_item, stream};
+
+    // Initialize symbol group id lookup table
+    SymbolGroupIdLookupT::InitDeviceSymbolGroupIdLut(sgid_init, symbol_vec, stream);
+
+    // Initialize state transition table
+    TransitionTableT::InitDeviceTransitionTable(transition_table_init, tt_vec, stream);
+
+    // Initialize finite-state transducer lookup table
+    TranslationTableT::InitDeviceTranslationTable(translation_table_init, out_tt_vec, stream);
+  }
+
+  template <typename SymbolT,
+            typename TransducedOutItT,
+            typename TransducedIndexOutItT,
+            typename TransducedCountOutItT,
+            typename OffsetT>
+  void Transduce(SymbolT const* d_chars,
```

Comment:
nit: Since this is the API used in `fst_test.cu`, it would nice to have some brief and param documentation here. (or use `@copybrief DeviceTranduce` or `@copydoc DeviceTransduce`)

Reply:
I've added documentation 👍